### DISCUSSION TITLE
Documentation Develop Processors

### DISCUSSION THREAD

**Discussion by flomickl** [apache/streampipes#1183] [Dev]
As mentioned in the community meeting
I added the documentation of what to know to write processors 

https://cwiki.apache.org/confluence/x/eJNbDg

There are plenty of blank spots:
e.g.

Data Example of a ListProperty
How to Read ListsProperties with Example
In general, Time in StreamPipes for example if merging Streams
Explain of all requiredInput Methods
Explanation of some output startegies and when to use.

I but some question in warning boxes.
Maybe we can discuss it here and fill it with even more content

**Comment by tenthe** [comment]
Cool, thanks for starting to collect all the options.
Did you also take a look at the "SKD" sections of the documentation? [1]
There are already some short descriptions there. Maybe we can expand them there.

We should discuss, where we want to have what kind of information. 
I like that you made a very detailed overview of all possibilities so that we can see how the current model looks like.
But I guess for users this will be overvelming and for us it will be hard to maintain when things change.

My suggestion would be to include this information in the Javadoc. In the SDK module [2] there is already a lot of JavaDoc, maybe we can check what is missing there and extend it.
All other information should be included in the documentation. There, I would especially like to have much more detailed descriptions of the concepts, as you have already started.

What do you think?

[1] https://streampipes.apache.org/docs/docs/extend-sdk-functions.html
[2] https://github.com/apache/streampipes/tree/dev/streampipes-sdk

**Reply by flomickl** [reply]
> Cool, thanks for starting to collect all the options. Did you also take a look at the "SKD" sections of the documentation? [1] There are already some short descriptions there. Maybe we can expand them there.


Oh I did not know this. Again, a new feature I missed? :-)

> 
 Yes, that was the plan, but for the start it was easier to create a separate doc to get a better overview. The main goal is to add all this information to the doc website in the end; in a more compact way. Maybe in smaller sections ?!?

> We should discuss, where we want to have what kind of information. I like that you made a very detailed overview of all possibilities so that we can see how the current model looks like. But I guess for users this will be overvelming and for us it will be hard to maintain when things change.
> 
> My suggestion would be to include this information in the Javadoc. In the SDK module [2] there is already a lot of JavaDoc, maybe we can check what is missing there and extend it. All other information should be included in the documentation. There, I would especially like to have much more detailed descriptions of the concepts, as you have already started.

Yes, Javadoc sound great. But in my opinion, there is never a good examples part in the Javadoc.

So I would suggest the description part is on the website doc and extend the SDK module or somewhere else?
All table sections could be referenced to the Javadoc. 
The examples should be again in the description section. 
If something big changed, you have to adjust the documentation anyway. But this way it is a good mixture.

What do you think?
Is there already a Javadoc?

> 
> What do you think?
> 
> [1] https://streampipes.apache.org/docs/docs/extend-sdk-functions.html [2] https://github.com/apache/streampipes/tree/dev/streampipes-sdk

**Reply by bossenti** [reply]
I think one midterm (or longterm) goal we need to accomblish is to render references from Javadocs as well and add them to our documentation. 
This could make it easier to connect both worlds and also doesn't require us to duplicate documentation at some parts.

**Reply by flomickl** [reply]
@bossenti @tenthe Is there already a Javadoc output somehow or is this a feature that needs to be generated?

**Reply by bossenti** [reply]
Are you referring to a rendered output, e.g. as HTML?
Not that I'm aware of (at least not for all code), but that's something I really would like to have as part of our StreamPipes docs

Maybe this would also help us to be more disciplined in terms of writing Java docs 😀
